test/test_simple_anchor.cpp

```diff
@@ -3,6 +3,410 @@
 namespace c4 {
 namespace yml {
 
+TEST(anchors, node_scalar_set_ref_when_empty)
+{
+    {
+        NodeScalar ns;
+        ns.set_ref_maybe_replacing_scalar("foo", /*has_scalar*/false);
+        EXPECT_EQ(ns.scalar, "foo");
+        EXPECT_EQ(ns.anchor, "foo");
+    }
+    {
+        NodeScalar ns;
+        ns.set_ref_maybe_replacing_scalar("*foo", /*has_scalar*/false);
+        EXPECT_EQ(ns.scalar, "*foo");
+        EXPECT_EQ(ns.anchor, "foo");
+    }
+}
+
+TEST(anchors, node_scalar_set_ref_when_non_empty)
+{
+    {
+        NodeScalar ns;
+        ns.scalar = "whatever";
+        ns.set_ref_maybe_replacing_scalar("foo", /*has_scalar*/true);
+        EXPECT_EQ(ns.scalar, "foo");
+        EXPECT_EQ(ns.anchor, "foo");
+    }
+    {
+        NodeScalar ns;
+        ns.scalar = "whatever";
+        ns.set_ref_maybe_replacing_scalar("*foo", /*has_scalar*/true);
+        EXPECT_EQ(ns.scalar, "*foo");
+        EXPECT_EQ(ns.anchor, "foo");
+        ns.set_ref_maybe_replacing_scalar("foo", /*has_scalar*/true);
+        EXPECT_EQ(ns.scalar, "*foo"); // keep the previous as it is well formed
+        EXPECT_EQ(ns.anchor, "foo");
+        ns.set_ref_maybe_replacing_scalar("bar", /*has_scalar*/true);
+        EXPECT_EQ(ns.scalar, "bar"); // replace previous as it is not well formed
+        EXPECT_EQ(ns.anchor, "bar");
+    }
+}
+
```

Comment:
@Gei0r I guess this PR is ready. The new behavior is consistent with what we previously discussed; but there are more nuances here, so take a look at the following tests to discover how it works, both from parsing and setting it programatically.

Reply:
Looks really nice, I'll give it a thorough try tomorrow.